src/Microsoft.OData.Client/DataServiceExtensions.cs

```diff
@@ -0,0 +1,156 @@
+                // In the event that a developer who adds a reference to the library invokes CountDistinct() as follows:
+                // - new List<int> { 1, 2, 1 }.AsQueryable().CountDistinct(d => d)
+
+                // Method: Select<TSource,TResult>(IQueryable<TSource>, Expression<Func<TSource,TResult>>)
+                MethodInfo selectMethod = GetSelectMethod();
+                // Method: Distinct<TSource>(IQueryable<TSource>)
+                MethodInfo distinctMethod = GetDistinctMethod(typeof(Queryable), typeof(IQueryable<>));
+                // Method: Count<TSource>(IQueryable<TSource>)
+                MethodInfo countMethod = GetCountMethod(typeof(Queryable), typeof(IQueryable<>));
+
+                // Select(d => d.Prop)
+                MethodCallExpression selectMethodExpr = Expression.Call(null,
+                    selectMethod.MakeGenericMethod(new Type[] { source.ElementType, selector.Body.Type }),
+                    new[] { source.Expression, Expression.Quote(selector) });
+
+                // Select(d => d.Prop).Distinct()
+                MethodCallExpression distinctMethodExpr = Expression.Call(null,
+                    distinctMethod.MakeGenericMethod(new Type[] { selector.Body.Type }),
+                    new[] { selectMethodExpr });
+
+                // Select(d => d.Prop).Distinct().Count()
+                countMethodExpr = Expression.Call(null,
+                    countMethod.MakeGenericMethod(new Type[] { selector.Body.Type }),
+                    new[] { distinctMethodExpr });
+            }
+
+            return source.Provider.Execute<int>(countMethodExpr);
+        }
+
+        /// <summary>
+        /// Returns the distinct count of elements in a sequence after applying the projection function to each element.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source"/></typeparam>
+        /// <typeparam name="TTarget">The type returned by the projection function represented in <paramref name="selector"/>.</typeparam>
+        /// <param name="source">A sequence of values of type <typeparamref name="TSource"/>.</param>
+        /// <param name="selector">A projection function to apply to each element.</param>
+        /// <returns>Distinct count of elements in a sequence after applying the projection function to each element.</returns>
+        public static int CountDistinct<TSource, TTarget>(this IEnumerable<TSource> source, Func<TSource, TTarget> selector)
+        {
+            // Provide a default implementation...
+            // In the event that a developer who adds a reference to the library invokes CountDistinct() as follows:
+            // - new List<int> { 1, 2, 1 }.CountDistinct(d => d)
+
+            // Method: Select<TSource,TResult>(IEnumerable<TSource>, Func<TSource,TResult>)
+            MethodInfo selectMethod = typeof(Enumerable).GetMethods(BindingFlags.Static | BindingFlags.Public)

```

Comment:
Is it possible to refractor the `GetSelectMethod` and move this logic there?

Reply:
The logic for getting the `Select` defined in `Enumerable` type differs a little from the one defined in `Queryable` type but I'll see how I can refactor and use distinct names for each.